snuba/snuba_migrations/events_analytics_platform/0002_spans_attributes_mv.py

```diff
@@ -0,0 +1,138 @@
+from __future__ import annotations
+
+from typing import Sequence
+
+from snuba.clickhouse.columns import AggregateFunction, Column, DateTime, String, UInt
+from snuba.clusters.storage_sets import StorageSetKey
+from snuba.migrations import migration, operations, table_engines
+from snuba.migrations.columns import MigrationModifiers as Modifiers
+from snuba.migrations.operations import OperationTarget, SqlOperation
+from snuba.utils.constants import ATTRIBUTE_BUCKETS
+
+META_KEY_QUERY_TEMPLATE = """
+SELECT
+    organization_id,
+    attribute_key,
+    {attribute_value} AS attribute_value,
+    toMonday(start_timestamp) AS timestamp,
+    retention_days,
+    sumState(cast(1, 'UInt64')) AS count
+FROM eap_spans_local
+LEFT ARRAY JOIN
+    arrayConcat({key_columns}) AS attribute_key,
+    arrayConcat({value_columns}) AS attr_value
+GROUP BY
+    organization_id,
+    attribute_key,
+    attribute_value,
+    timestamp,
+    retention_days
+"""
+
+
+class Migration(migration.ClickhouseNodeMigration):
+    """
+    This migration creates a table meant to store just the attributes seen in a particular org.
+    The table is populated by a separate materialized view for each type of attribute.
+    """
+
+    blocking = False
+    storage_set_key = StorageSetKey.EVENTS_ANALYTICS_PLATFORM
+    granularity = "8192"
+
+    value_types = ["str", "num"]
+
+    meta_view_name = "spans_attributes_{attribute_type}_meta_mv"
+    meta_local_table_name = "spans_attributes_meta_local"
+    meta_dist_table_name = "spans_attributes_meta_dist"
+    meta_table_columns: Sequence[Column[Modifiers]] = [
+        Column("organization_id", UInt(64)),
+        Column("attribute_type", String()),
+        Column("attribute_key", String()),
+        Column("attribute_value", String()),
+        Column("timestamp", DateTime(modifiers=Modifiers(codecs=["DoubleDelta"]))),
+        Column("retention_days", UInt(16)),
+        Column("count", AggregateFunction("sum", [UInt(64)])),
+    ]
+
+    def forwards_ops(self) -> Sequence[SqlOperation]:
+        create_table_ops: list[SqlOperation] = [
+            operations.CreateTable(
+                storage_set=self.storage_set_key,
+                table_name=self.meta_local_table_name,
+                engine=table_engines.AggregatingMergeTree(
+                    storage_set=self.storage_set_key,
+                    primary_key="(organization_id, attribute_key)",
+                    order_by="(organization_id, attribute_key, attribute_value, timestamp)",
```

Comment:
if timestamp is in the sorting key, nothing will ever be aggregated? what's the point of using aggregating at all?

Reply:
The timestamp is rounded to the nearest Monday in the query, so it will aggregate an entire week of data.

Not sure if that answers your question at all.